diff --git a/source/hostap-2.11/src/ap/wpa_auth_ie.c b/source/hostap-2.11/src/ap/wpa_auth_ie.c
index 5908f07c4..c3faa3f5c 100644
--- a/source/hostap-2.11/src/ap/wpa_auth_ie.c
+++ b/source/hostap-2.11/src/ap/wpa_auth_ie.c
@@ -874,9 +1321,15 @@ wpa_validate_wpa_ie(struct wpa_authenticator *wpa_auth,
 	    wpa_auth->conf.sae_require_mfp &&
 	    wpa_key_mgmt_sae(sm->wpa_key_mgmt) &&
 	    !(data.capabilities & WPA_CAPABILITY_MFPC)) {
-		wpa_printf(MSG_DEBUG,
-			   "Management frame protection required with SAE, but client did not enable it");
-		return WPA_MGMT_FRAME_PROTECTION_VIOLATION;
+		if (key_mgmt & WPA_KEY_MGMT_PSK) {
+			wpa_printf(MSG_DEBUG,
+					"Management frame protection required with SAE, but client supports PSK");
+			sm->wpa_key_mgmt = WPA_KEY_MGMT_PSK;
+		} else {
+			wpa_printf(MSG_DEBUG,
+					"Management frame protection required with SAE, but client did not enable it");
+			return WPA_MGMT_FRAME_PROTECTION_VIOLATION;
+		}
 	}
 #endif /* CONFIG_SAE */
